File: services/core/entity/goal_input.go
package entity

import (
	"time"

	"tenkhours/pkg/types"
)

type GoalInput struct {
	ID          *string           `json:"id,omitempty"`
	Name        string            `json:"name"                  validate:"min=1,max=50"`
	Description *string           `json:"description,omitempty" validate:"omitempty,max=255"`
	StartTime   time.Time         `json:"startTime"`
	EndTime     time.Time         `json:"endTime"`
	Metrics     []GoalMetricInput `json:"metrics"`
	Checkboxes  []CheckboxInput   `json:"checkboxes"`
}

type GoalMetricInput struct {
	ID          string          `json:"id"`
	Condition   MetricCondition `json:"condition"`
	TargetValue *float64        `json:"targetValue"`
	RangeValue  *RangeInput     `json:"rangeValue,omitempty"`
}

type RangeInput struct {
	Min float64 `json:"min"`
	Max float64 `json:"max"`
}

type CheckboxInput struct {
	ID    *string `json:"id"`
	Name  string  `json:"name"`
	Value bool    `json:"value"`
}

type MetricCondition string

const (
	MetricConditionGreaterThan      MetricCondition = "gt"
	MetricConditionLessThan         MetricCondition = "lt"
	MetricConditionEqual            MetricCondition = "eq"
	MetricConditionLessThanEqual    MetricCondition = "lte"
	MetricConditionGreaterThanEqual MetricCondition = "gte"
	MetricConditionInRange          MetricCondition = "ir"
)

type GoalPipeline struct {
	Filter  *GoalFilter
	OrderBy *GoalOrderBy
	*types.Pagination
}

type GoalFilter struct {
	CharacterID *string     `json:"character_id"`
	Status      *GoalStatus `json:"status"`
}

type GoalOrderBy struct{}

File: services/penk/src/utils/grpc/proto/core/GoalInput.ts
// Original file: ../../proto/core/goal_message.proto

import type { GoalMetricInput as _core_GoalMetricInput, GoalMetricInput__Output as _core_GoalMetricInput__Output } from '../core/GoalMetricInput';
import type { CheckboxInput as _core_CheckboxInput, CheckboxInput__Output as _core_CheckboxInput__Output } from '../core/CheckboxInput';
import type { Long } from '@grpc/proto-loader';

export interface GoalInput {
  'id'?: (string);
  'characterId'?: (string);
  'name'?: (string);
  'description'?: (string);
  'startTime'?: (number | string | Long);
  'endTime'?: (number | string | Long);
  'metrics'?: (_core_GoalMetricInput)[];
  'checkboxes'?: (_core_CheckboxInput)[];
  '_id'?: "id";
  '_description'?: "description";
}

export interface GoalInput__Output {
  'id'?: (string);
  'characterId': (string);
  'name': (string);
  'description'?: (string);
  'startTime': (string);
  'endTime': (string);
  'metrics': (_core_GoalMetricInput__Output)[];
  'checkboxes': (_core_CheckboxInput__Output)[];
  '_id': "id";
  '_description': "description";
}
